### PR TITLE
Metadata with complete type information

### LINKED ISSUE #917: Fully self-descriptive metadata types
One thing I am missing has to do with the Types one can find in the Metadata. 

Some types are (Rust?!) base types such as ‘u32’, ‘bool’ or base structures such as ‘vec’. It is quite clear how to deal with these in an objective manner.

What I am missing has to do with the higher level objects such as T::Balance, ValidatorPrefs, Address, et cetera. The metadata is not self-descriptive enough to ‘know’ the composition of these higher level objects/structures. Hence an outside system (such as a 'generalized substrate explorer' [https://polkascan.io]) would need to ‘hard code’ the composition of such types or structures.

So I suppose what i am suggesting is to extend the metadata with a description of the composition of higher level objects and structures.

PS: discussion with Gavin on Riot hinted that this is  non-trivial change and will likely not be picked up before 2020. Hence this issue is for future reference.

### PR DESCRIPTION
**Pull request by xlc**
This fixes #917, which is one of the biggest pain point to me when working with Substrate (and polkadot.js), that any type information needs to be implemented twice and incompatible type information simply cause polkadot.js crashes without much helpful clues (other than it is mostly likely a codec error).

This introduce `type_registry` to `RuntimeMetadata` which contains all the type informations i.e. the metadata itself stores the information about how to decode the storage types.

Ideally, this will remove the needs to manually maintain type information in polkadot.js  (currently implemented at https://github.com/polkadot-js/api/tree/master/packages/types/src). Those JS classes can be generated at runtime by parsing the metadata. TS files can be generated ahead-of-time if needed. This will also make implement (and support) other languages SDK much easily. 

Some code copied from parity-codec.

Fixes #917
Enable the proper fix for https://github.com/polkadot-js/api/issues/416
Could be the foundation work for https://github.com/polkadot-js/api/issues/429

This is the debug print of the resulting metadata:
https://gist.github.com/xlc/19d010e55f84bf6d43afafdc5e5b5df6#file-metadata-L5525

### CODE REVIEW

**Comment by bkchr** [comment]
Nice and ty. That was something I was thinking to do as well. Good that you started it :)
The main problem with your implementation is the blow up of human readable static stuff in the binary. In the first implementation of the metadata, we used json to represent it, but we switched to binary to get rid of all static strings lying in the runtime and increasing its size. So I would not accept your current implementation.
As I thought about it, I came to the idea of exposing fundamental types as a fixed binary representation. So, `u64` could be `b1` and `u32` could be `b01`. Limiting this to 256 different values should be enough. The procedural macro could use a fixed list to generate the correct binary representation per type. And if a type is not supported, we would see this at compilation and could prepare an update of the Javascript side as well.

**Comment by xlc** [comment]
I can remove the `name` field and change `TypeMetadata::Primative` to `TypeMetadata::Primative(u8)`, and possibility remove enum variant name and field name which should solve the concern of having human readable string in the wasm file.
However this means it can be harder to generate a developer friendly sdk / typescript file from it because there will be no more names. Maybe use some `std` feature flag to still have a verbose mode for native compilation?

**Comment by bkchr** [comment]
First, you would just have a map:
```
BinaryToType = {
   0 => u64,
   1 -> u32,
    ...
}
```

For adding a new type, you would just need to add `2 -> i32`. The javascript side could just say, values above `1` are unknown.
I think that if we support all fundamental types, there will be no missing types at all. Even if you would encode them as string, how does the javascript side knows how to interpret them?
Everything will be `u`someting to `i`something. Or a key that uses x bits.
As a team will not introduce any new fundamental types to rust, the javascript side should be fairly stable, with a good initial set of known types.

**Comment by xlc** [comment]
@bkchr I have updated it to use enum (u8) to encode primitive types instead of string.
There are more optimization could be done, such that currently if a complex type (struct/enum) is used is multiple places, it will be encoded multiple times. But those are not important for PoC purpose.

Please let me know if you think this is the correct direction to go and the best way to structure the repos. This substrate-metadata module to me can be either part of parity-codec, or a separate lib depends on parity-codec. And some method name is a bit awkward (`type_metadat` method was originally called `metadata`, which conflicts do some existing `metadata` method so I have to rename it), it will be good if we can come up some better and less confusing names.

I want to keep working on this but I don't want to send too much time on it and then found out it towards to a dead end.

**Comment by gavofyork** [comment]
Looks pretty cool overall (though it'll need to work with the `Compact` encoding stuff)

**Comment by xlc** [comment]
**Needs help**
I tried multiple ways but still not able to make it work with `HasCompact`. My Rust skill is still limited and I have trouble understand some implementation details of parity-codec and why `HasCompact` is implemented as it is.

Some methods I tried:
Modify parity-codec `HasCompact` trait, but upgrade `parity-codec` to master was causing some unexpected compiling error

```
error[E0277]: the trait bound `<<T as Trait>::Moment as codec::HasCompact>::Type: std::convert::From<<T as Trait>::Moment>` is not satisfied
   --> srml/timestamp/src/lib.rs:162:56
    |
162 |         vec![(T::TIMESTAMP_SET_POSITION, Call::set(next_time.into()))]
    |                                                              ^^^^ the trait `std::convert::From<<T as Trait>::Moment>` is not implemented for `<<T as Trait>::Moment as codec::HasCompact>::Type`
    |
    = help: consider adding a `where <<T as Trait>::Moment as codec::HasCompact>::Type: std::convert::From<<T as Trait>::Moment>` bound
    = note: required because of the requirements on the impl of `std::convert::Into<<<T as Trait>::Moment as codec::HasCompact>::Type>` for `<T as Trait>::Moment`
```

https://github.com/xlc/substrate/commit/7ac594fc4ff29d7d3bafd1661b7b81d66df28c1f

I tried duplicate the implementation of `HasCompact` in `substrate-metadata` module and change the runtime modules to use the duplicated and modified one, but still getting all compiling error that make me I wonder why the initial version was able to compile at all.

I tried replace `<T as HasCompac>::Type` to `Compac<T>` and still having many unexpected compiling issues.

I am also thinking move the metadata code into parity-codec, which should simplify some implementation details. Do you guys think is this something part of parity-codec?

I am stuck and need help to continue.

**Comment by xlc** [comment]
https://github.com/paritytech/substrate/pull/1499 Unblocked my issue, but https://github.com/paritytech/substrate/pull/1473 makes circular type dependency so I will need to make some refactor to handle it

**Comment by xlc** [comment]
This is PoC ready and updated output: https://gist.github.com/xlc/0d25d0af60b3e4a01570b78926c66902#file-substrate-runtime-metadata-L865

It is still unoptimized and very inefficient but enough to serve the purpose as PoC.

This is going to conflict with #1428 so I will wait until it is merged and then update this branch.

Please provide feedbacks.

**Comment by gavofyork** [comment]
I think you might need to rephrase your question.

File: core/metadata/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,413 @@
...
+			"u32" => PrimativeMetadata::U32,
+			"i32" => PrimativeMetadata::I32,
+			"u64" => PrimativeMetadata::U64,
+			"i64" => PrimativeMetadata::I64,
+			"u128" => PrimativeMetadata::U128,
+			"i128" => PrimativeMetadata::I128,
+			_ => PrimativeMetadata::Unknown, // panic! can make it runtime error but how to make this compile-time error?
+		}
+	}
+}
```
Comment (thiolliere) [inline core/metadata/src/lib.rs]: It seems to me you can implement it only for those types:
```rust
macro_rules! impl_primatives {
	( $( $t:ty => $p:expr),* ) => { $(
		impl EncodeMetadata for $t {
			fn type_name() -> MetadataName {
				vec![stringify!($t).into()]
			}

			fn type_metadata_kind(_registry: &mut MetadataRegistry) -> TypeMetadataKind {
				TypeMetadataKind::Primative($p)
			}
		}
	)* }
}

impl_primatives!(
	u8 => PrimativeMetadata::U8,
	u16 => PrimativeMetadata::U16,
	u32 => PrimativeMetadata::U32,
	u64 => PrimativeMetadata::U64,
	u128 => PrimativeMetadata::U128,
	usize => PrimativeMetadata::Usize,
	i8 => PrimativeMetadata::I8,
	i16 => PrimativeMetadata::I16,
	i32 => PrimativeMetadata::I32,
	i64 => PrimativeMetadata::I64,
	i128 => PrimativeMetadata::I128,
	isize => PrimativeMetadata::Isize,
	bool => PrimativeMetadata::Bool
);
```

**Comment by thiolliere** [comment]
I'm finally less confused

* first: question how do you generate this https://gist.github.com/xlc/0d25d0af60b3e4a01570b78926c66902#file-substrate-runtime-metadata-L865

* second: why don't we use something like:
```rust
pub enum PrimativeMetadata {
	Unknown,
	Unit,
	PhantomData, // do we need this or it can just be Unit?
	Bool,
	Usize, Isize,
	U8, I8,
	U16, I16,
	U32, I32,
	U64, I64,
	U128, I128,
	Array(u32, Box<PrimativeMetadata>),
	Vector(Box<PrimativeMetadata>),
	Tuple(Vec<Box<PrimativeMetadata>>),
	Struct(Vec<(FieldName, Box<PrimativeMetadata>)>),
	...
}
```
it seems to me more efficient as we don't have to store all names.

the output would look more like:
```rust
ty: "T::Hash",
type_metadata: MetadataRegistry {
	list: [
		TypeMetadata::Array(32, TypeMetadata::Primative::u8)),
	]
}
```
instead of 
```rust
                            ty: "T::Hash",
                            type_metadata: MetadataRegistry {
                                list: [
                                    TypeMetadata {
                                        kind: Array( 32, [ "u8" ]),
                                        name: [ "Array", "32" ]
                                    },
                                    TypeMetadata {
                                        kind: Primative( U8),
                                        name: [ "u8" ]
                                    }
                                ]
                            }
```

**Comment by xlc** [comment]
@thiolliere Thanks for the feedbacks.
1) I generate the metadata by update the code to debug print it in native build.
2) Yes I think it is good idea to flatten the structure

Some kind of name / reference / id is needed to handle recursive types (e.g. A contains B, which contains Option<A>). I need a way to see if a type is already registered, and reference it by name / id. But yes I think I can avoid have names for primitives .

**Comment by thiolliere** [comment]
actually recursive type seems ok if you use Box. 
Probably having primative in its own enum would allow to be encoded in a smaller way.
Also using name would be interesting if we don't want to define many times the same type.
But in the current implementation the following enums seems to work.
```rust
#[derive(Encode, Clone, PartialEq, Eq)]
#[cfg_attr(feature = "std", derive(Decode, Debug, Serialize))]
pub enum PrimativeMetadata {
	Unknown,
	Unit,
	PhantomData, // do we need this or it can just be Unit?
	Bool,
	Usize, Isize,
	U8, I8,
	U16, I16,
	U32, I32,
	U64, I64,
	U128, I128,
}

#[derive(Encode, Clone, PartialEq, Eq)]
#[cfg_attr(feature = "std", derive(Decode, Debug, Serialize))]
pub enum TypeMetadataKind {
	Primative(PrimativeMetadata),
	Array(u32, Box<TypeMetadataKind>),
	Vector(Box<TypeMetadataKind>),
	Struct(Vec<(StringBuf, Box<TypeMetadataKind>)>),
	Enum(Vec<..>),
	Tuple(Vec<..>),
	Compact,
}
```

**Comment by rphmeier** [comment]
btw, it is spelled `Primitive`, not `Primative`

**Comment by xlc** [comment]
This is mostly working now. Just needs fix conflicts and tests

Latest metadata: https://gist.github.com/xlc/3778f339afb9c1f4557ac9391cd14928#file-gistfile1-txt-L4290

**Comment by xlc** [comment]
This is ready for review.

Can someone have a look why this build is failing on CI? It works on my machine...
```
$ cargo -V
cargo 1.33.0-nightly (2b4a5f1f0 2019-01-12)
```

**Comment by thiolliere** [comment]
The CI use rust stable. maybe you used a coming feature for path:
```

error[E0433]: failed to resolve: could not find `dispatch` in `{{root}}`
   --> srml/support/src/dispatch.rs:885:34
    |
885 |         #[derive(Clone, PartialEq, Eq, $crate::substrate_metadata_derive::EncodeMetadata)]
    |                                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ could not find `dispatch` in `{{root}}`
```

**Comment by xlc** [comment]
@thiolliere Thanks for look it. What should be the fix for this? Remove `$crate::substrate_metadata_derive::` part will throws compile error on the crates that does not have the `EncodeMetadata` macro in scope. Maybe I can fix each of them but it doesn't seems ideal.

**Comment by thiolliere** [comment]
I have hard time understanding why it is complaining: if I remove `$crate::substrate_metadata_derive::` and launch `cargo test -p srml-support` I got the same error that it couldn't find dispatch.

I can't find any solution, even non-ideal ones. I don't know at all why $crate doesn't work here, Is it related to the use of the procedural macro ?

**Comment by xlc** [comment]
@thiolliere I removed the nightly feature and have everything compiles and run fine with rust nightly. I also identified the code causing the compile error on rust stable
https://github.com/paritytech/substrate/pull/1328/commits/548b5340238e5a6071d30120951b3faff9fe0376
But still don't understand what is happening here... Potentially a rust compiler issue?

File: srml/assets/src/lib.rs [hunk 2]
```diff
@@ -22,6 +22,7 @@
 use srml_support::{StorageValue, StorageMap, Parameter, decl_module, decl_event, decl_storage, ensure};
 use primitives::traits::{Member, SimpleArithmetic, Zero, StaticLookup};
 use system::ensure_signed;
+use substrate_metadata_derive::EncodeMetadata;
```
Comment (cheme) [inline srml/assets/src/lib.rs]: I tried this kind of code (as an alternative to the $crate in derive issue), seems ok but I did not test further, it might be a dead end or totally unrelated, just adding this comment in case it can help.
```diff
diff --git a/Cargo.lock b/Cargo.lock
index 63d7c26c..e0aaec65 100644
Binary files a/Cargo.lock and b/Cargo.lock differ
diff --git a/node/runtime/wasm/target/wasm32-unknown-unknown/release/node_runtime.compact.wasm b/node/runtime/wasm/target/wasm32-unknown-unknown/release/node_runtime.compact.wasm
index 33545160..2cf8a858 100644
Binary files a/node/runtime/wasm/target/wasm32-unknown-unknown/release/node_runtime.compact.wasm and b/node/runtime/wasm/target/wasm32-unknown-unknown/release/node_runtime.compact.wasm differ
diff --git a/srml/assets/Cargo.toml b/srml/assets/Cargo.toml
index 4d5d30b7..cca4c078 100644
--- a/srml/assets/Cargo.toml
+++ b/srml/assets/Cargo.toml
@@ -15,7 +15,7 @@ srml-support = { path = "../support", default-features = false }
 # `system` module provides us with all sorts of useful stuff and macros depend on it being around.
 system = { package = "srml-system", path = "../system", default-features = false }
 substrate-metadata = { path = "../../core/metadata", default-features = false }
-substrate-metadata-derive = { path = "../../core/metadata-derive", default-features = false }
+#substrate-metadata-derive = { path = "../../core/metadata-derive", default-features = false }
 
 [dev-dependencies]
 substrate-primitives = { path = "../../core/primitives", default-features = false }
@@ -31,5 +31,5 @@ std = [
 	"srml-support/std",
 	"system/std",
 	"substrate-metadata/std",
-	"substrate-metadata-derive/std",
+#	"substrate-metadata-derive/std",
 ]
diff --git a/srml/assets/src/lib.rs b/srml/assets/src/lib.rs
index 39014ad6..beda3d9c 100644
--- a/srml/assets/src/lib.rs
+++ b/srml/assets/src/lib.rs
@@ -22,7 +22,7 @@
 use srml_support::{StorageValue, StorageMap, Parameter, decl_module, decl_event, decl_storage, ensure};
 use primitives::traits::{Member, SimpleArithmetic, Zero, StaticLookup};
 use system::ensure_signed;
-use substrate_metadata_derive::EncodeMetadata;
+//use substrate_metadata_derive::EncodeMetadata;
 
 pub trait Trait: system::Trait {
 	/// The overarching event type.
diff --git a/srml/support/src/dispatch.rs b/srml/support/src/dispatch.rs
index b67a8959..143a0716 100644
--- a/srml/support/src/dispatch.rs
+++ b/srml/support/src/dispatch.rs
@@ -605,6 +605,9 @@ macro_rules! decl_module {
 			}
 		)*
 
+    mod dummy_import{
+      use super::*;
+	use $crate::substrate_metadata_derive::EncodeMetadata;
 		#[cfg(feature = "std")]
 		$(#[$attr])*
 		#[derive(EncodeMetadata)]
@@ -628,7 +631,8 @@ macro_rules! decl_module {
 				$fn_name ( $( $param ),* ),
 			)*
 		}
-
+    }
+    pub use dummy_import::*;
 		// manual implementation of clone/eq/partialeq because using derive erroneously requires
 		// clone/eq/partialeq from T.
 		impl<$trait_instance: $trait_name> $crate::dispatch::Clone
```

Comment (xlc) [reply srml/assets/src/lib.rs]: thanks for the suggestion. this is definitely cleaner

**Comment by bkchr** [comment]
Hey @xlc, I tried to reproduce this error:

> error[E0433]: failed to resolve: could not find `dispatch` in `{{root}}`
>    --> srml/support/src/dispatch.rs:885:34
>     |
> 885 |         #[derive(Clone, PartialEq, Eq, $crate::substrate_metadata_derive::EncodeMetadata)]
>     |                                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ could not find `dispatch` in `{{root}}`

But I could not reproduce it with a stable compiler and @thiolliere could not reproduce it anymore as well. Maybe you pulled in some other changes with one of your rebases? Nevertheless, I would really not want to create some static module with `decl_module!`. Can you try to bring back your old solution with the full path to `EncodeMetadata` in `derive`. If it is working for you locally with a stable compiler as well and CI is the failing point, we will look into making the CI happy.

**Comment by xlc** [comment]
@bkchr revert this https://github.com/paritytech/substrate/pull/1328/commits/548b5340238e5a6071d30120951b3faff9fe0376 and you have the error. I don't know what's happening.

The solution with `$crate::substrate_metadata_derive::EncodeMetadata` works for me with a nightly rust version. I can't get it working with stable version.

The solution proposed by cheme above seems to be working and I am exploring it.

**Comment by thiolliere** [comment]
can I ask for your rust version just in case ? we used last stable `rustc 1.32.0 (9fda7c223 2019-01-16)`

**Comment by bkchr** [comment]
I reverted `548b5340238e5a6071d30120951b3faff9fe0376` and it still works. Maybe you fetched some other fix with one of your rebases?

**Comment by xlc** [comment]
I am using `rustc 1.32.0 (9fda7c223 2019-01-16)` and with 88815c3ec8ece7eb36cca3fae6ace9516e59b108

```
$ cargo check --tests -p srml-support
    Checking srml-support v0.1.0 (/Users/bryanchen/projects/cennznet/substrate/srml/support)
error[E0433]: failed to resolve: could not find `dispatch` in `{{root}}`
   --> srml/support/src/dispatch.rs:898:35
    |
898 |             #[derive(Clone, PartialEq, Eq, EncodeMetadata)]
    |                                            ^^^^^^^^^^^^^^ could not find `dispatch` in `{{root}}`
```

**Comment by bkchr** [comment]
Okay, @thiolliere and I debugged it. What is happening can be seen here in the following playground: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=042752cb4704ecff80a374b5380ed8ae

We tested with beta as well and there the bug is fixed as well as in nightly. So, just revert all your test hacks etc and bring back `$crate::substrate_metadata_derive::EncodeMetadata`. We will merge this pr after the next rust version is released (should be end of February).

**Comment by xlc** [comment]
Thanks for the help. So it is actually a compiler bug... I have reverted the code. I will continue testing this with a nightly rust for now.

**Comment by xlc** [comment]
@bkchr I am not sure why [this](c32a2ab19cdc055f5492b184f80ffed7b8accc64) is needed but otherwise it gives me those error
```
error[E0433]: failed to resolve: could not find `srml_support` in `{{root}}`
 --> <::support::dispatch::impl_outer_dispatch macros>:7:26
  |
7 |   Clone , PartialEq , Eq , $ crate :: substrate_metadata_derive ::
  |  __________________________^
8 | | EncodeMetadata ) ] # [ cfg_attr ( feature = "std" , derive ( Debug ) ) ] pub
  | |______________^ could not find `srml_support` in `{{root}}`

    Checking substrate-consensus-aura v0.1.0 (/Users/bryanchen/projects/cennznet/substrate/core/consensus/aura)
    Checking node-template-runtime v0.9.0 (/Users/bryanchen/projects/cennznet/substrate/node-template/runtime)
error: aborting due to previous error

For more information about this error, try `rustc --explain E0433`.
error: Could not compile `node-runtime`.
warning: build failed, waiting for other jobs to finish...
error[E0433]: failed to resolve: could not find `srml_support` in `{{root}}`
 --> <::srml_support::dispatch::impl_outer_dispatch macros>:7:26
  |
7 |   Clone , PartialEq , Eq , $ crate :: substrate_metadata_derive ::
  |  __________________________^
8 | | EncodeMetadata ) ] # [ cfg_attr ( feature = "std" , derive ( Debug ) ) ] pub
  | |______________^ could not find `srml_support` in `{{root}}`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0433`.
error: Could not compile `node-template-runtime`.
warning: build failed, waiting for other jobs to finish...
error[E0433]: failed to resolve: could not find `srml_support` in `{{root}}`
 --> <::support::dispatch::impl_outer_dispatch macros>:7:26
  |
7 |   Clone , PartialEq , Eq , $ crate :: substrate_metadata_derive ::
  |  __________________________^
8 | | EncodeMetadata ) ] # [ cfg_attr ( feature = "std" , derive ( Debug ) ) ] pub
  | |______________^ could not find `srml_support` in `{{root}}`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0433`.
error: Could not compile `node-runtime`.

To learn more, run the command again with --verbose.
```

**Comment by xlc** [comment]
This still have one limitation that it doesn't allow flexible metadata formats. i.e. Modify metadata format is still a breaking change.
I am going to make the metadata itself includes additional metadata type information so that modify of metadata format can still be handled. Of cause there is some limits and some types are just cannot be changed without introduce a breaking change (bumping metadata version)

**Comment by bkchr** [comment]
> @bkchr I am not sure why [c32a2ab](https://github.com/paritytech/substrate/commit/c32a2ab19cdc055f5492b184f80ffed7b8accc64) is needed but otherwise it gives me those error
> 
> ```
> error[E0433]: failed to resolve: could not find `srml_support` in `{{root}}`
>  --> <::support::dispatch::impl_outer_dispatch macros>:7:26
>   |
> 7 |   Clone , PartialEq , Eq , $ crate :: substrate_metadata_derive ::
>   |  __________________________^
> 8 | | EncodeMetadata ) ] # [ cfg_attr ( feature = "std" , derive ( Debug ) ) ] pub
>   | |______________^ could not find `srml_support` in `{{root}}`
> 
>     Checking substrate-consensus-aura v0.1.0 (/Users/bryanchen/projects/cennznet/substrate/core/consensus/aura)
>     Checking node-template-runtime v0.9.0 (/Users/bryanchen/projects/cennznet/substrate/node-template/runtime)
> error: aborting due to previous error
> 
> For more information about this error, try `rustc --explain E0433`.
> error: Could not compile `node-runtime`.
> warning: build failed, waiting for other jobs to finish...
> error[E0433]: failed to resolve: could not find `srml_support` in `{{root}}`
>  --> <::srml_support::dispatch::impl_outer_dispatch macros>:7:26
>   |
> 7 |   Clone , PartialEq , Eq , $ crate :: substrate_metadata_derive ::
>   |  __________________________^
> 8 | | EncodeMetadata ) ] # [ cfg_attr ( feature = "std" , derive ( Debug ) ) ] pub
>   | |______________^ could not find `srml_support` in `{{root}}`
> 
> error: aborting due to previous error
> 
> For more information about this error, try `rustc --explain E0433`.
> error: Could not compile `node-template-runtime`.
> warning: build failed, waiting for other jobs to finish...
> error[E0433]: failed to resolve: could not find `srml_support` in `{{root}}`
>  --> <::support::dispatch::impl_outer_dispatch macros>:7:26
>   |
> 7 |   Clone , PartialEq , Eq , $ crate :: substrate_metadata_derive ::
>   |  __________________________^
> 8 | | EncodeMetadata ) ] # [ cfg_attr ( feature = "std" , derive ( Debug ) ) ] pub
>   | |______________^ could not find `srml_support` in `{{root}}`
> 
> error: aborting due to previous error
> 
> For more information about this error, try `rustc --explain E0433`.
> error: Could not compile `node-runtime`.
> 
> To learn more, run the command again with --verbose.
> ```
> 
> using `rustc 1.33.0-nightly (ceb251214 2019-01-16)`

Weird, but you will need to look into the expanded source code what is happening. We don't want people to include crates by a fixed name.

**Comment by xlc** [comment]
@bkchr I can't just expand it because it will fail with the same error. I did some hack to make EncodeMetadata macro to generate the code wrapped in stringify to avoid the compile error.

https://gist.github.com/xlc/7608d67d67bc8133a5f61bab85d85503#file-expanded-rs-L1405

I really don't understand why it fails. `::srml_support` are referenced in other parts of code and they don't cause issues. Only in the code generated by EncodeMetadata cannot contain it.

**Comment by xlc** [comment]
I did some more experiment and code generated from master is using `$crate::dispatch` whereas the code generated from this branch is using `::srml_support::dispatch`.
Any Rust config I failed to merge into this branch or something else?

e.g.
```
pub enum Call {
    Timestamp($crate::dispatch::CallableCallFor<Timestamp>),
    Consensus($crate::dispatch::CallableCallFor<Consensus>),
    Indices($crate::dispatch::CallableCallFor<Indices>),
    Balances($crate::dispatch::CallableCallFor<Balances>),
    Session($crate::dispatch::CallableCallFor<Session>),
    Staking($crate::dispatch::CallableCallFor<Staking>),
    Democracy($crate::dispatch::CallableCallFor<Democracy>),
    Council($crate::dispatch::CallableCallFor<Council>),
    CouncilVoting($crate::dispatch::CallableCallFor<CouncilVoting>),
    CouncilMotions($crate::dispatch::CallableCallFor<CouncilMotions>),
    Grandpa($crate::dispatch::CallableCallFor<Grandpa>),
    Treasury($crate::dispatch::CallableCallFor<Treasury>),
    Contract($crate::dispatch::CallableCallFor<Contract>),
    Sudo($crate::dispatch::CallableCallFor<Sudo>),
}
```

vs 

```
pub enum Call {
    Timestamp(::srml_support::dispatch::CallableCallFor<Timestamp>),
    Consensus(::srml_support::dispatch::CallableCallFor<Consensus>),
    Indices(::srml_support::dispatch::CallableCallFor<Indices>),
    Balances(::srml_support::dispatch::CallableCallFor<Balances>),
    Session(::srml_support::dispatch::CallableCallFor<Session>),
    Staking(::srml_support::dispatch::CallableCallFor<Staking>),
    Democracy(::srml_support::dispatch::CallableCallFor<Democracy>),
    Council(::srml_support::dispatch::CallableCallFor<Council>),
    CouncilVoting(::srml_support::dispatch::CallableCallFor<CouncilVoting>),
    CouncilMotions(::srml_support::dispatch::CallableCallFor<CouncilMotions>),
    Grandpa(::srml_support::dispatch::CallableCallFor<Grandpa>),
    Treasury(::srml_support::dispatch::CallableCallFor<Treasury>),
    Contract(::srml_support::dispatch::CallableCallFor<Contract>),
    Sudo(::srml_support::dispatch::CallableCallFor<Sudo>),
}
```

**Comment by xlc** [comment]
I have workaround the issue by manually implement EcodeMetadata for Call to avoid using derive. Seems like a Rust bug to me.

**Comment by xlc** [comment]
@bkchr @jacogr I have made the `AccountId` a new type instead of `H256` because there are some [special code on the polkadot.js](https://github.com/polkadot-js/common/blob/98e2964c73f02cd49e91dc654d9eb8795b8d8dd1/packages/keyring/src/address/encode.ts) for base58 multi base encoding/decoding. It is hard to kept existing behavior without telling polkadot.js which H256 is AccountId and which is plain hash type. On the other hand, I don't wish to kept the the previous behavior that using type alias name because it is not something can be relied on. So I figured that the new type approach is most appropriate here.

This is the commit of the change (still have some compiler issue on Wasm build)
https://github.com/paritytech/substrate/pull/1328/commits/6d12e43a3e88372f69933f26bc63116d9c657ed7

Let me know if you think there is a better way to handle it.

cc @ianhe8x

**Comment by xlc** [comment]
@jacogr One issue we have found out is that [`Vec<KeyValue>` here](https://gist.github.com/xlc/3778f339afb9c1f4557ac9391cd14928#file-gistfile1-txt-L512) is actually `Vec<(Vec<u8>,Vec<u8>)>` in metadata V2.
It is [`Vec<KeyValue>`](https://github.com/polkadot-js/api/blob/b7fa5e9f28e8b1b7f8895b39e7f63a53fc3b0b80/packages/types/src/type/KeyValue.ts#L37)  in metadata V1 and have some mismatch in polkadot.js codec (while having codec compatibility so it works).

How should we deal with this mismatch issue? Change it on substrate side make a proper `struct KeyValue { key: Vec<u8>, value: Vec<u8> }`, or just break polkadot.js?

**Comment by xlc** [comment]
Regards to https://github.com/polkadot-js/api/pull/712#issuecomment-465910202, I have three options, but I don't think I can make the decision myself so please help.
1. Keep old metadata name. e.g. the old metadata type name `BalanceOf<T>` will be kept, and can be used by SDK to aid customization. But this comes with major drawback that it will be a maintenance burden to ensure SDK is able to make sense of the metadata name used by substrate, which should be implementation detail. But now any type called `Balance` will have a special meaning which could cause some surprises to parachain developers.
2. Add annotation to types, similar to `#[compac]`, to provide addition context to SDK. This is a reasonable approach to me but increase the overhead to implement runtime methods as every `Balance` like fields should be annotated with `#[metadata(Balance)]` in order to get the `Balance` behavior on SDK side.
3. Use new type approach, by change `type Balance = u128` to `struct Balance(u128)`. This is my preferred way.

**Comment by jacogr** [comment]
@xlc So if you look at KeyValue (which actually is treated more or less as a primitive on the polkadot-js API side), it was _specifically_ done that way since the KeyValue has special meaning and properties comes to interpreting. So polkadot-js is using the fact there that `(vec<u8>,vec<u8>)` is equivalent to` struct { k: vec<u8>, v: vec<u8> }` for encoding. (It could very well have been a Tuple with the same `key` and `value` properties exposed)

(Since it is used relatively intensively, just a generic Tuple with no properties for API-specific use does not do the trick since it would be error-prone expansion all-over.)

Substrate should not cater for idiosyncrasies.

**Comment by xlc** [comment]
I get it. But I also feel this `KeyValue` special case handling should be stay in SDK side, instead of substrate side. That means we needs a way to tell polkadot.js that the parameter for `consensus.set_storage` should be replaced with `Vec<KeyValue>`. But this have its drawbacks as well.
I need a decision but I don't think I have the authority to make one. So someone please make one.

**Comment by jacogr** [comment]
100% agreed - any kind of special interpretation (read: making it easy for higher level interactions) belong on the SDK side. As long as we carry the type name as user (like in the case of Balance vs u128) and the breakdown to primitives it is workable to “enhance” on the SDK side.

**Comment by xlc** [comment]
But in terms of this PR implementation, which option do you prefer? I feel you want to do (1), i.e. keep the original string names?

**Comment by jacogr** [comment]
So on the options 1-4.

4. This loses metadata richness. I've raised the issues that I see in polkadot-js API and dependent projects with it, cannot speak for other implementations, but believe some similar issues probably exist. Dumbing down everything to "just" primitives without identifying information is a step backwards.

3. This has been done in some places to aid codec work. It is simple and works. Drawbacks are that it does add burden to maintenance, e.g. "why is it this way" and "can I add extra fields since it is a struct".

2. This works... until it doesn't and some area is forgotten about for decoration. Once again I believe it increases the maintenance burden, i.e. things are done in a specific way to cater for downstream.

1. Unless I misunderstand, this works. Name the same, the metadata extended type info still indicates how this maps to the right type of primitive. (e.g. since the info is there, no need to maintain/define type classes to have these known)

TL;DR - I believe keeping the V1 views (names there exactly the same, call/event/storage definitions the same) and adding a mapping to the underlying types ensures compatibilty (among decoders/encoders). and removes the burden of maintaining these type mappings for injection by 3rd parties and type definitions by the various SDK layers. (polkadot-js, oo7, Python - the 3 I know of)

**Comment by xlc** [comment]
I can definitely do options 1 and I do agree this is a quick solution to implement and make sense to do it.

My concerns are this creates a hidden dependency between the type name used in substrate, and SDK the interoperation of the SDK implementation. It is very possible some other parachain developer add a new `Address` type, and use it to be represent a street address, and cause unwanted side effects on SDK side. This also means a refactor on substrate side may break SDK side.

As long as we all fully understand the consequence the decision, I have no opposition.

**Comment by xlc** [comment]
Latest metadata with options 1
https://gist.github.com/xlc/19d010e55f84bf6d43afafdc5e5b5df6#file-metadata-L5525

File: core/metadata-derive/src/encode.rs [hunk 3]
```diff
@@ -0,0 +1,206 @@
+use quote::{quote, quote_spanned};
```
Comment (DemiMarie-parity) [inline core/metadata-derive/src/encode.rs]: I have no experience writing procedural macros.  @rphmeier can you help review this?

File: core/sr-primitives/src/generic/unchecked_mortal_compact_extrinsic.rs [hunk 4]
```diff
@@ -186,12 +188,14 @@ impl<Address, Index, Call, Signature> fmt::Debug for UncheckedMortalCompactExtri
 	}
 }
 
+#[cfg(test)]
```
Comment (DemiMarie-parity) [inline core/sr-primitives/src/generic/unchecked_mortal_compact_extrinsic.rs]: is this intentional?

Comment (xlc) [reply core/sr-primitives/src/generic/unchecked_mortal_compact_extrinsic.rs]: no. probably a merge error

File: core/metadata-derive/src/encode.rs (hunk 3 above)
Comment (rphmeier) [reply core/metadata-derive/src/encode.rs]: I don't have much experience with proc-macros or this code in particular either.

**Comment by siman** [comment]
No automatically generated type metadata anymore?

**Comment by xlc** [comment]
We still plan to do it but unlikely within next two weeks.

**Comment by folsen** [comment]
Just a note here, the closing is not because we don't want the feature, but because the PR had gotten stale enough that it's easier to re-open fresh when ready than trying to resolve all merge conflicts, we regularly close stale issues and PRs because of uncertainty around status, so again, not an indication of desire to have it or not.

**Comment by siman** [comment]
Great!
